Added device placement tutorial.

regrettable-username opened the pull request:
This PR introduces the following:
- A new Jupyter Notebook that walks through how to use the device placement APIs.
- Changes to the `README.md` to reference the aforementioned notebook.
- Edit to `_book.yaml` to reference the new notebook from the documentation site.

regrettable-username commented:
I signed it!

regrettable-username commented:
This PR addresses https://github.com/tensorflow/swift/issues/167

rxwei commented on docs/site/tutorials/device_placement.ipynb:
```diff
@@ -0,0 +1,378 @@
+{
```
<p>The use of a `get` prefix in function names is generally discouraged in Swift. The Swift API Design Guidelines state:</p><blockquote>Name functions and methods according to their side-effect.</blockquote><p><br></p><p>`get` should be removed from these names.</p><p><br></p><p>For more details on naming, see <a href="https://swift.org/documentation/api-design-guidelines/#promote-clear-usage" target="_blank">Swift API Design Guidelines - Promote Clear Usage</a>.</p><p><br></p><p><br></p><p>Moreover, the `getTimeString` function can be renamed to be more clear and not to repeat the return type information. And the body of the `getTimeString` function can be improved to use early return, or even better, pattern matching.</p><p><br></p><p>```swift</p><p>func timeDescription(_ nanoseconds: Double) -&gt; String {</p><p>switch nanoseconds {</p><p>case ..&lt;1e3:</p><p>return "\(nanoseconds) ns"</p><p>case ..&lt;1e6:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;return "\(nanoseconds/1e3) µs"</p><p>case ..&lt;1e9:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;return "\(nanoseconds/1e6) ms"</p><p>default:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;return "\(nanoseconds/1e9) seconds"</p><p>}</p><p>}</p><p>```</p><p><br></p><p><br></p><p>In your first function, please avoid the use of&nbsp;explicit `Array&lt;T&gt;`. Also, prefer type inference and literal conversion over explicit calls to the initializer.</p><p>```swift</p><p>var timings: [Double] = 0.0</p><p>```</p><p>For more details on types with short hand names like array and dictionary, see the <a href="https://google.github.io/swift/#types-with-shorthand-names" target="_blank">Google Swift Style Guide</a>.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/168/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

regrettable-username replied:
Ah, of course! This should be corrected now.

regrettable-username replied:
Got it.

rxwei commented on docs/site/tutorials/device_placement.ipynb:
```diff
@@ -0,0 +1,378 @@
+{
```
<p>For consistency with other Swift for TensorFlow tutorials, please only capitalize the first letter in a sentence in headings.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/168/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

rxwei replied:
<p>For consistency with other tutorials, use bare form verbs instead of gerunds in headings and titles.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/168/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

regrettable-username commented:
I believe I've addressed all the feedback. Please let me know if there's anything else. In the event that things look good, I'll want to update the README to point to the notebook on `tensorflow/swift` instead of my fork. I'm also curious if the edits to `_book.yaml` look okay. I'm not sure how the tensorflow.org documentation is generated, but I just tried to keep it consistent with the other colab notebook tutorials. Thanks.

rxwei commented:
Sorry for the delay in getting back to you. And thanks a lot for your work! We are currently busy with v0.3 release. I should be able to re-review and merge it next week.

regrettable-username commented:
Hey, no worries. I'm sure you guys are super busy over there. I just now realized I had used snake case in a few spots and have corrected it. I guess I've been writing too much Python recently :sweat_smile:

marcrasi commented on docs/site/tutorials/device_placement.ipynb:
```diff
@@ -0,0 +1,372 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This tutorial demonstrates how to use the device placement APIs in Swift for TensorFlow. Device placement APIs allow you to run operations on a specific device, for example, on a CPU or a GPU."
```
Could you add a sentence that says this? "The current device placement APIs are particularly experimental and might change."

marcrasi commented:
Closing for now because of lack of activity, but feel free to reopen if you have time to take another look!
